### ISSUE TITLE
Development Environment in Dashboards

### ISSUE THREAD

**Issue by nicovon24** [thingsboard/thingsboard#10781]
Hello, 
I am encountering an issue within my current company regarding the absence of a development environment in Thingsboard. Unlike my previous experiences in various courses and companies where I had the opportunity to work with development environments, Thingsboard lacks this crucial feature.

My current workflow involves creating both production and development dashboards. However, this approach is inefficient and prone to errors. I find myself having to manually transfer changes from the development to the production environment, often resulting in oversights or missed updates in the process.

This inefficiency was recently highlighted when presenting the application developed with Thingsboard to a respected tester. He expressed dissatisfaction with our approach, particularly our use of client dashboards in the production environment.

Given that popular platforms like GitHub and many open-source projects provide development environments, I strongly believe that Thingsboard should offer a similar feature. Such an environment would greatly streamline our workflow, ensuring smoother transitions from development to production and less errors with our clients.

In addition to that, it would also provide a more modern software development environment and more people would be interessed in the platform, we have more than 700 devices and 20 really relevant dashboards and an error in production could cost my company a client, in a thing which could be detected in development and testing side

Thanks!

**Comment by devaskim** [comment]
Hi @nicovon24 
In general, I agree with you that it is very serious questions. However it is not clear at least for me what you want to get from the TB development environment. So maybe try to describe your ideas in details

**Comment by nicovon24** [comment]
Hello! Currently a dashboard is in development until you save it, after you save it, it is in production. This had lead us usually to problems with clients.
I would suggest having branches like in github in the dashboards, I dont know if it is possible
I understand that currently is only in the backup recovery in the platform only with Github, but would be nice to have in the dashboards. Have a development side and production side.
thanks!

**Comment by jrkubic** [comment]
I agree that this described development environment would be very helpful. As a workaround, we created two AWS EC2 instances and ran a Thingsboard instance on each. Following a bit of this [development flow](https://deepwiki.com/thingsboard/thingsboard-extensions/2.2-development-workflow)

<img width="468" height="296" alt="Image" src="https://github.com/user-attachments/assets/08955dd1-3a49-4e0d-8fa7-7c3d2ce32504" />

Basically, you build as much as you can in custom widgets, then build them into [extensions](https://github.com/thingsboard/thingsboard-extensions)

Finally, take the whole flow and implement with a Gitlab pipeline.

Unless you package the entire dash into the extension, you will still be left with some nuances. For instance, some TB oriented stuff is pretty tough to bring into the extension (ie rule chains).

But this was helpful for major widget and UI work, possibly a solution for what you need until a formal TB Dev environment is made.

**Comment by mykhailokornieiev** [comment]
Hi. Thank you for the suggestion. Such a feature request was previously added to our backlog for developers. 
We will review this request and will get back to you with updates as soon as they arise. 
Please follow our [Release Notes](https://thingsboard.io/docs/reference/releases/) to see all new additions and fixes to the platform in subsequent releases.

**Comment by mykhailokornieiev** [comment]
Hi. We've created an Idea discussion https://github.com/thingsboard/thingsboard/discussions/15781 so that we can check the community's opinion regarding this request. This is done to identify the most desirable features for the community with further analysis of this data and its transfer to developers for further review. Even if the feature request is rejected, this information will still be useful for further planning the platform's development, and the decision may be reconsidered in the future.
